conductr_cli/conduct_backup.py

```diff
@@ -0,0 +1,262 @@
+        self.args = args
+        self.bundle_id_or_name = args.bundle
+
+    def backup_conductr(self):
+        log = logging.getLogger(__name__)
+
+        backup_path, backup_name = create_backup_directory(self.args.output_path)
+        try:
+            initial_bundles_raw = bundles(self.args)
+            if initial_bundles_raw is not False:
+                initial_bundles_json = json.loads(initial_bundles_raw)
+                initial_bundle_core_info = BundleCoreInfo.from_bundles(initial_bundles_json)
+
+                if self.bundle_id_or_name is None:
+
+                    for info in initial_bundle_core_info:
+                        self.backup_bundle(backup_path, info)
+
+                    final_bundles_raw = bundles(self.args)
+                    final_bundles_json = json.loads(final_bundles_raw)
+                    final_bundle_core_info = BundleCoreInfo.from_bundles(final_bundles_json)
+
+                    self.process_diff(backup_path, initial_bundle_core_info, final_bundle_core_info)
+                    self.backup_bundle_json(backup_path, final_bundles_raw)
+
+                else:
+                    bundle_info = BundleCoreInfo.filter_by_bundle_id(initial_bundle_core_info, self.bundle_id_or_name)
+
+                    if bundle_info is None:
+                        raise ConductBackUpError(str.format("Bundle {0} was not found.", self.bundle_id_or_name))
+
+                    filtered_bundle = filter_bundles_by_id_or_name(initial_bundles_json, self.bundle_id_or_name)
+                    self.backup_bundle_json(backup_path, json.dumps(filtered_bundle))
+                    self.backup_bundle(backup_path, bundle_info)
+
+                self.backup_members(backup_path)
+                self.backup_agents(backup_path)
+                compressed_backup_path = self.compress_backup(self.args.output_path, backup_path, backup_name)
+                remove_backup_directory(backup_path)
+
+                log.info("Backup process was successful.")
+                log.info("Backup was written to : " + compressed_backup_path)
+
+        except ConductBackUpError as error:
+            log.error(error.messages)
+            log.error("Rolling back all temp files.")
+            remove_backup_directory(backup_path)
+            raise error
+
+    def process_diff(self, backup_path, initial_bundle_core_info, final_bundle_core_info):
+        bundles_removed = BundleCoreInfo.diff(initial_bundle_core_info, final_bundle_core_info)
+        bundles_added = BundleCoreInfo.diff(final_bundle_core_info, initial_bundle_core_info)
+        for bundle_info in bundles_removed:
+            self.remove_bundle(backup_path, bundle_info)
+
+        for bundle_info in bundles_added:
+            self.backup_bundle(backup_path, bundle_info)
+
+    def backup_bundle(self, backup_path, bundle_core_info):
+        log = logging.getLogger(__name__)
+        log.info("Processing bundle : " + bundle_core_info.bundle_id)
```

Comment:
Maybe make this `log.debug`? I'm not sure. I am just considering the scripting aspect I was talking about and it would be nice as an operator to do that without grepping.